cub/test/catch2_test_device_radix_sort_keys.cu

```diff
@@ -0,0 +1,471 @@
+              thrust::raw_pointer_cast(out_keys.data()),
+              num_items,
+              begin_bit<key_t>(),
+              end_bit<key_t>());
+  }
+
+  // Perform a bitwise comparison to ensure that 0 != -0:
+  REQUIRE_BITWISE_EQ(ref_keys, out_keys);
+}
+
+CUB_TEST("DeviceRadixSort::SortKeys: NaN handling", "[keys][radix][sort][device]", fp_key_types)
+{
+  using key_t    = c2h::get<0, TestType>;
+  using limits_t = cuda::std::numeric_limits<key_t>;
+
+  const std::size_t num_items = (1 << 18) + 7;
+  thrust::device_vector<key_t> in_keys(num_items);
+  thrust::device_vector<key_t> out_keys(num_items);
+
+  const int num_key_seeds = 1;
+  c2h::gen(CUB_SEED(num_key_seeds), in_keys);
+
+  // Sprinkle some NaNs randomly throughout the keys:
+  {
+    const size_t num_indices = num_items / 128;
+    thrust::device_vector<std::size_t> indices(num_indices);
+    bool has_nans = false;
+    for (int i = 0; i < 2; ++i)
+    {
+      const bool supported = i == 0 ? limits_t::has_signaling_NaN : limits_t::has_quiet_NaN;
+      const key_t nan_val  = i == 0 ? limits_t::signaling_NaN() : limits_t::quiet_NaN();
+
+      if (supported)
+      {
+        has_nans = true;
+        c2h::gen(CUB_SEED(1), indices, std::size_t(0), num_items);
+        auto begin = thrust::make_constant_iterator(nan_val);
+        auto end   = begin + num_indices;
+        thrust::scatter(begin, end, indices.cbegin(), in_keys.begin());
+      }
+    }
+    if (!has_nans)
+    {
+      // SKIP(); Not available until Catch2 3.3.0
+      return;
+    }
+  }
+
+  const bool is_descending = GENERATE(false, true);
+
+  auto ref_keys = radix_sort_reference(in_keys, is_descending);
+
+  if (is_descending)
+  {
+    sort_keys_descending(
+      thrust::raw_pointer_cast(in_keys.data()),
+      thrust::raw_pointer_cast(out_keys.data()),
+      num_items,
+      begin_bit<key_t>(),
+      end_bit<key_t>());
+  }
+  else
+  {
+    sort_keys(thrust::raw_pointer_cast(in_keys.data()),
+              thrust::raw_pointer_cast(out_keys.data()),
+              num_items,
+              begin_bit<key_t>(),
+              end_bit<key_t>());
+  }
+
+  REQUIRE_EQ_WITH_NAN_MATCHING(ref_keys, out_keys);
+}
+
+CUB_TEST("DeviceRadixSort::SortKeys: entropy reduction", "[keys][radix][sort][device]", single_key_type)
+{
+  using key_t = c2h::get<0, TestType>;
+
+  const std::size_t num_items = (1 << 18) + 7;
+  thrust::device_vector<key_t> in_keys(num_items);
+
+  const int num_key_seeds = 1;
+  c2h::gen(CUB_SEED(num_key_seeds), in_keys);
+
+  // Repeatedly bitwise-and random keys together. This increases the likelyhood
+  // of duplicate keys.
+  const int entropy_reduction = GENERATE(1, 3, 9, 15);
+  {
+    thrust::device_vector<key_t> tmp(num_items);
+    for (int i = 0; i < entropy_reduction; ++i)
+    {
+      c2h::gen(CUB_SEED(1), tmp);
+      thrust::transform(in_keys.cbegin(), in_keys.cend(), tmp.cbegin(), in_keys.begin(), thrust::bit_and<key_t>{});
+    }
+  }
```

Comment:
question: is this functionally equivalent to using `c2h::gen` with reduced min / max range?

Reply:
No -- this will still test the full range of the datatype, allowing all bits to potentially be set while reducing the overall number of set bits.